### ISSUE TITLE
hpcrun: no or small samples for cache-references and cache-misses events

### ISSUE THREAD

**Issue by laksono** [HPCToolkit/hpctoolkit#70]
Profiling a program with -e cache-references and  -e cache-misses generates very small samples and sometimes no samples at all.

Platforms tested: ori, les, knl (all are x86)

```
/home/la5/pkgs/hpctoolkit-kernel/bin/hpcrun -dd LINUX_PERF -e cache-misses -e cache-references /home/la5/pkgs/hpctoolkit-kernel/bin/hpcstruct   /home/la5/pkgs/hpctoolkit-kernel/libexec/hpctoolkit/hpcstruct-bin
...
$ grep samples hpctoolkit-hpcstruct-bin-measurements/*.log
SUMMARY: samples: 23 (recorded: 23, blocked: 0, errant: 0, trolled: 0, yielded: 0)
```
With hpcproftt:
```
[metric-tbl: (num-entries: 2)
  [(nm: cache-misses) (desc: cache-misses) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 0) (period-mean: 0.000000) (num-samples: 0)]
  [(nm: cache-references) (desc: cache-references) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 0) (period-mean: 66518.826087) (num-samples: 23)]
]
```
Linux perf can generate many samples (around 20k samples). There must be a problem with the initialization of perf events.
Interestingly, if we force to use PAPI sample-source, PAPI doesn't recognize the events.

**Comment by laksono** [comment]
fixed in commit be7ebd5b99da1f4851cab1a4ac01dbd9413563da

**Comment by laksono** [comment]
It fails to detect precise_ip correctly on Cori
Linux cori03 4.4.103-92.56-default #1 SMP Wed Dec 27 16:24:31 UTC 2017 (2fd2155) x86_64 x86_64 x86_64 GNU/Linux
Inside the log file, it states that the value of precise_ip is 2
```
[75824, N]: LINUX_PERF: event fd: 9, skid: 2, code: 0, type: 0, period: 249, freq: 1
[75824, N]: LINUX_PERF: event fd: 10, skid: 2, code: 1, type: 0, period: 249, freq: 1
[75824, N]: LINUX_PERF: event fd: 16, skid: 2, code: 2, type: 0, period: 249, freq: 1
[75824, N]: LINUX_PERF: event fd: 17, skid: 2, code: 3, type: 0, period: 249, freq: 1
[75824, N]: LINUX_PERF: event fd: 18, skid: 3, code: 2, type: 1, period: 249, freq: 1
```
hpcprofttt shows that there is no samples for cache events:
```
  [(nm: CYCLES) (desc: CYCLES) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 1) (period-mean: 12240173.782154) (num-samples: 177)]
  [(nm: INSTRUCTIONS) (desc: INSTRUCTIONS) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 1) (period-mean: 3209281.605166) (num-samples: 97)]
  [(nm: cache-references) (desc: cache-references) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 0) (period-mean: 0.000000) (num-samples: 0)]
  [(nm: cache-misses) (desc: cache-misses) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 0) (period-mean: 0.000000) (num-samples: 0)]
  [(nm: PAGE-FAULTS) (desc: PAGE-FAULTS) ((ty: 1) (val-ty: 0) (val-fmt: 2) (partner: 0) (show: 1) (showPercent: 1)) (period: 1) (formula: ) (format: )
    (frequency: 1) (multiplexed: 0) (period-mean: 51.758621) (num-samples: 29)]
```

**Comment by laksono** [comment]
Temporarily fixed in commit baa162d2c5fa960275b48848e7b296c3ca1afd2b
By default, we set precise_ip to zero, unless users specify the HPCRUN_PRECISE_IP variable.

We should enable users to force precise events by adding a modifier just like perf tool, so each event is unique.
